### PR TITLE
ICU-21656 fix wrong matching categories.

### PR DESCRIPTION
**Pull request by younies**
closes: https://github.com/icu-units/icu/issues/130

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21656
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [ ] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by younies** [comment]
@sffc , I have used the convertibility instead of simplifying because of the following reasons:

1. If the unit is reciprocal (such as in consumption) it should match
2. it is very hard to simplify: square-meter-per-millimeter.

**Comment by sffc** [comment]
As discussed in our weekly meeting:

(1) This approach works, but it is fragile to locale data changes and could be slow

(2) The units we are simplifying have no power prefixes

(3) Mark suggested the following algorithm instead:

1. Check if the base unit has an exact match
2. Check if the reciprocal of the base unit has an exact match
3. Check if the fully simplified base unit has an exact match
4. Check if the reciprocal of the fully simplified base unit has an exact match
5. If still no matches, return an error code.

**Comment by younies** [comment]
@sffc , I have used the new methodology that we have discussed in the meeting and I have added two new tests for unit consumption.

If the code is good for you, I will start porting it to C++.

**Comment by younies** [comment]
@sffc , now everything is there, ready for final review

**Review by sffc — COMMENTED** [review]
Almost there!

File: icu4c/source/i18n/measunit_extra.cpp [hunk 1]
```diff
@@ -989,6 +1027,33 @@ void MeasureUnitImpl::takeReciprocal(UErrorCode& /*status*/) {
...
+                break;
+            }
+        }
+
+        if (!unitExist) {
+            result.appendSingleUnit(singleUnit, status);
+        }
+    }
+
+    return result;
```
[the remark this answers is not in the document]
Comment (younies) [reply icu4c/source/i18n/measunit_extra.cpp:1054]: First of all, we normalize the units with their SI prefix, so
 - square-kilogram-per-kilogram --> kilogram
 - square-kilogram-per-gram --> square-kilogram-per-gram


Secondly, when we build the `MeasureUnitImpl` to get the Id, we sort the units.

**Review by sffc — COMMENTED** [review]
LGTM but please fix the build failure reported by CI:

```
   g++	 ...  ../../../icu4c/source/test/intltest/numbertest_decimalquantity.cpp
../../../icu4c/source/test/intltest/numbertest_api.cpp: In member function ‘void NumberFormatterApiTest::unitUsage()’:
../../../icu4c/source/test/intltest/numbertest_api.cpp:1798:47: error: expected ‘)’ before ‘unit’
 1798 |                        NumberFormatter::with()                                     //
      |                                               ^
      |                                               )
 1799 |                        unit(MeasureUnit::forIdentifier("gallon-per-mile", status)) //
      |                        ~~~~                    
../../../icu4c/source/test/intltest/numbertest_api.cpp:1795:23: note: to match this ‘(’
 1795 |     assertFormatSingle(u"gallon-per-mile match the correct category",              //
      |                       ^
```

**Review by sffc — CHANGES_REQUESTED** [review]
1. Please don't force-push until the end, because it makes it harder to track changes between reviews.
2. The ASAN/Valgrind build is failing; please fix.

**Comment by younies** [comment]
@sffc , now `asan` is solved, the problem was that when we build the `MeasureUnitImpl`, we need the following:

1. make a copy of `MeasureUnitImpl`, so, it can be used later.
2. we need to keep the `MeasureUnit` after we extract the `identifier` or copy the `identifier`


Here is a problem, to keep the identifier, you need to keep the `MeasureUnit`. I am not sure if returning `const char *` for `getIdentifier()` is a good choice.

File: icu4c/source/i18n/measunit_extra.cpp [hunk 2]
```diff
@@ -989,6 +1027,33 @@ void MeasureUnitImpl::takeReciprocal(UErrorCode& /*status*/) {
     }
 }
 
+MeasureUnitImpl MeasureUnitImpl::simplify(UErrorCode &status) const {
```
[the remark this answers is not in the document]
Comment (younies) [reply icu4c/source/i18n/measunit_extra.cpp:1036]: done!

File: icu4c/source/i18n/measunit_extra.cpp [hunk 3]
```diff
@@ -847,29 +840,80 @@ umeas_getPrefixBase(UMeasurePrefix unitPrefix) {
...
+    UErrorCode localStatus = U_ZERO_ERROR;
+    umtx_initOnce(gUnitExtrasInitOnce, &initUnitExtras, status);
     if (U_FAILURE(status)) {
         return result;
     }
-    UErrorCode localStatus = U_ZERO_ERROR;
-    int32_t idx = getUnitCategoryIndex(baseUnitIdentifier, localStatus);
+    BytesTrie trie(gSerializedUnitCategoriesTrie);
+    MeasureUnit baseMeasureUnit = std::move(baseMeasureUnitImpl.copy(status)).build(status);
+    StringPiece identifier = baseMeasureUnit.getIdentifier();
```
[the remark this answers is not in the document]
Comment (younies) [reply icu4c/source/i18n/measunit_extra.cpp]: done!

File: icu4c/source/i18n/measunit_extra.cpp [hunk 4]
```diff
@@ -847,29 +840,80 @@ umeas_getPrefixBase(UMeasurePrefix unitPrefix) {
...
-        // fallback logic: inversion and fallback may depend on presence or
-        // absence of a usage for that category.
-        if (uprv_strcmp(baseUnitIdentifier.data(), "meter-per-cubic-meter") == 0) {
-            result.append(kConsumption, (int32_t)kConsumptionLen, status);
+        localStatus = U_ZERO_ERROR;
+        MeasureUnitImpl baseMeasureUnitReciprocal = baseMeasureUnitImpl.copy(status);
+        baseMeasureUnitReciprocal.takeReciprocal(status);
+        identifier.clear();
+        MeasureUnit reciprocal = std::move(baseMeasureUnitReciprocal).build(status);
+        identifier.set(reciprocal.getIdentifier());
```
[the remark this answers is not in the document]
Comment (younies) [reply icu4c/source/i18n/measunit_extra.cpp]: done!

File: icu4c/source/i18n/measunit_extra.cpp [hunk 5]
```diff
@@ -1113,7 +1184,7 @@ void MeasureUnitImpl::serialize(UErrorCode &status) {
 
 MeasureUnit MeasureUnitImpl::build(UErrorCode& status) && {
     this->serialize(status);
-    return MeasureUnit(std::move(*this));
+    return MeasureUnit(this->copy(status));
 }
```
[the remark this answers is not in the document]
Comment (younies) [reply icu4c/source/i18n/measunit_extra.cpp]: done!

File: icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitsData.java [hunk 6]
```diff
@@ -68,18 +69,39 @@ public static int getCategoryIndexOfSimpleUnit(int simpleUnitIndex) {
      * @return the corresponding category.
      */
     public String getCategory(MeasureUnitImpl measureUnit) {
-        MeasureUnitImpl baseMeasureUnit
+        MeasureUnitImpl baseMeasureUnitImpl
                 = this.getConversionRates().extractCompoundBaseUnit(measureUnit);
-        String baseUnitIdentifier = MeasureUnit.fromMeasureUnitImpl(baseMeasureUnit).getIdentifier();
+        String identifier = MeasureUnit.fromMeasureUnitImpl(baseMeasureUnitImpl).getIdentifier();
```
[the remark this answers is not in the document]
Comment (younies) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitsData.java]: done!

**Review by sffc — APPROVED** [review]
LGTM! Please squash

**Review by sffc — APPROVED** [review]

**Pull request merged**